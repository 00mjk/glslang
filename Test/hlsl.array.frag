float4 a[4];

struct {
    float4 m[7];
} s[11];

static float4 C = float4(1,2,3,4);

float4 PixelShaderFunction(int i : sem1, float4 input[3] : sem2) : SV_TARGET0
{
    float4 b[10] = { C, C, C, C, C, C, C, C, C, C };
    return a[1] + a[i] + input[2] + input[i] + b[5] + b[i] + s[i].m[i];
}
