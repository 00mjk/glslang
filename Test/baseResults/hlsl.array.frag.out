hlsl.array.frag
Shader version: 500
gl_FragCoord origin is upper left
0:? Sequence
0:7  Sequence
0:7    move second child to first child ( temp 4-component vector of float)
0:7      'C' ( global 4-component vector of float)
0:?       Constant:
0:?         1.000000
0:?         2.000000
0:?         3.000000
0:?         4.000000
0:10  Function Definition: @PixelShaderFunction(i1;vf4[3]; ( temp 4-component vector of float)
0:10    Function Parameters: 
0:10      'i' ( in int)
0:10      'input' ( in 3-element array of 4-component vector of float)
0:?     Sequence
0:11      Sequence
0:11        move second child to first child ( temp 10-element array of 4-component vector of float)
0:11          'b' ( temp 10-element array of 4-component vector of float)
0:11          Construct vec4 ( temp 10-element array of 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:12      Branch: Return with expression
0:12        add ( temp 4-component vector of float)
0:12          add ( temp 4-component vector of float)
0:12            add ( temp 4-component vector of float)
0:12              add ( temp 4-component vector of float)
0:12                add ( temp 4-component vector of float)
0:12                  add ( temp 4-component vector of float)
0:12                    direct index ( temp 4-component vector of float)
0:12                      a: direct index for structure ( uniform 4-element array of 4-component vector of float)
0:12                        'anon@0' (layout( row_major std140) uniform block{ uniform 4-element array of 4-component vector of float a,  uniform 11-element array of structure{ temp 7-element array of 4-component vector of float m} s})
0:12                        Constant:
0:12                          0 (const uint)
0:12                      Constant:
0:12                        1 (const int)
0:12                    indirect index ( temp 4-component vector of float)
0:12                      a: direct index for structure ( uniform 4-element array of 4-component vector of float)
0:12                        'anon@0' (layout( row_major std140) uniform block{ uniform 4-element array of 4-component vector of float a,  uniform 11-element array of structure{ temp 7-element array of 4-component vector of float m} s})
0:12                        Constant:
0:12                          0 (const uint)
0:12                      'i' ( in int)
0:12                  direct index ( temp 4-component vector of float)
0:12                    'input' ( in 3-element array of 4-component vector of float)
0:12                    Constant:
0:12                      2 (const int)
0:12                indirect index ( temp 4-component vector of float)
0:12                  'input' ( in 3-element array of 4-component vector of float)
0:12                  'i' ( in int)
0:12              direct index ( temp 4-component vector of float)
0:12                'b' ( temp 10-element array of 4-component vector of float)
0:12                Constant:
0:12                  5 (const int)
0:12            indirect index ( temp 4-component vector of float)
0:12              'b' ( temp 10-element array of 4-component vector of float)
0:12              'i' ( in int)
0:12          indirect index ( temp 4-component vector of float)
0:12            m: direct index for structure ( temp 7-element array of 4-component vector of float)
0:12              indirect index ( temp structure{ temp 7-element array of 4-component vector of float m})
0:12                s: direct index for structure ( uniform 11-element array of structure{ temp 7-element array of 4-component vector of float m})
0:12                  'anon@0' (layout( row_major std140) uniform block{ uniform 4-element array of 4-component vector of float a,  uniform 11-element array of structure{ temp 7-element array of 4-component vector of float m} s})
0:12                  Constant:
0:12                    1 (const uint)
0:12                'i' ( in int)
0:12              Constant:
0:12                0 (const int)
0:12            'i' ( in int)
0:10  Function Definition: PixelShaderFunction( ( temp void)
0:10    Function Parameters: 
0:?     Sequence
0:10      move second child to first child ( temp int)
0:?         'i' ( temp int)
0:?         'i' (layout( location=0) flat in int)
0:10      move second child to first child ( temp 3-element array of 4-component vector of float)
0:?         'input' ( temp 3-element array of 4-component vector of float)
0:?         'input' (layout( location=1) in 3-element array of 4-component vector of float)
0:10      move second child to first child ( temp 4-component vector of float)
0:?         '@entryPointOutput' (layout( location=0) out 4-component vector of float)
0:10        Function Call: @PixelShaderFunction(i1;vf4[3]; ( temp 4-component vector of float)
0:?           'i' ( temp int)
0:?           'input' ( temp 3-element array of 4-component vector of float)
0:?   Linker Objects
0:?     'anon@0' (layout( row_major std140) uniform block{ uniform 4-element array of 4-component vector of float a,  uniform 11-element array of structure{ temp 7-element array of 4-component vector of float m} s})
0:?     'C' ( global 4-component vector of float)
0:?     '@entryPointOutput' (layout( location=0) out 4-component vector of float)
0:?     'i' (layout( location=0) flat in int)
0:?     'input' (layout( location=1) in 3-element array of 4-component vector of float)


Linked fragment stage:


Shader version: 500
gl_FragCoord origin is upper left
0:? Sequence
0:7  Sequence
0:7    move second child to first child ( temp 4-component vector of float)
0:7      'C' ( global 4-component vector of float)
0:?       Constant:
0:?         1.000000
0:?         2.000000
0:?         3.000000
0:?         4.000000
0:10  Function Definition: @PixelShaderFunction(i1;vf4[3]; ( temp 4-component vector of float)
0:10    Function Parameters: 
0:10      'i' ( in int)
0:10      'input' ( in 3-element array of 4-component vector of float)
0:?     Sequence
0:11      Sequence
0:11        move second child to first child ( temp 10-element array of 4-component vector of float)
0:11          'b' ( temp 10-element array of 4-component vector of float)
0:11          Construct vec4 ( temp 10-element array of 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:11            'C' ( global 4-component vector of float)
0:12      Branch: Return with expression
0:12        add ( temp 4-component vector of float)
0:12          add ( temp 4-component vector of float)
0:12            add ( temp 4-component vector of float)
0:12              add ( temp 4-component vector of float)
0:12                add ( temp 4-component vector of float)
0:12                  add ( temp 4-component vector of float)
0:12                    direct index ( temp 4-component vector of float)
0:12                      a: direct index for structure ( uniform 4-element array of 4-component vector of float)
0:12                        'anon@0' (layout( row_major std140) uniform block{ uniform 4-element array of 4-component vector of float a,  uniform 11-element array of structure{ temp 7-element array of 4-component vector of float m} s})
0:12                        Constant:
0:12                          0 (const uint)
0:12                      Constant:
0:12                        1 (const int)
0:12                    indirect index ( temp 4-component vector of float)
0:12                      a: direct index for structure ( uniform 4-element array of 4-component vector of float)
0:12                        'anon@0' (layout( row_major std140) uniform block{ uniform 4-element array of 4-component vector of float a,  uniform 11-element array of structure{ temp 7-element array of 4-component vector of float m} s})
0:12                        Constant:
0:12                          0 (const uint)
0:12                      'i' ( in int)
0:12                  direct index ( temp 4-component vector of float)
0:12                    'input' ( in 3-element array of 4-component vector of float)
0:12                    Constant:
0:12                      2 (const int)
0:12                indirect index ( temp 4-component vector of float)
0:12                  'input' ( in 3-element array of 4-component vector of float)
0:12                  'i' ( in int)
0:12              direct index ( temp 4-component vector of float)
0:12                'b' ( temp 10-element array of 4-component vector of float)
0:12                Constant:
0:12                  5 (const int)
0:12            indirect index ( temp 4-component vector of float)
0:12              'b' ( temp 10-element array of 4-component vector of float)
0:12              'i' ( in int)
0:12          indirect index ( temp 4-component vector of float)
0:12            m: direct index for structure ( temp 7-element array of 4-component vector of float)
0:12              indirect index ( temp structure{ temp 7-element array of 4-component vector of float m})
0:12                s: direct index for structure ( uniform 11-element array of structure{ temp 7-element array of 4-component vector of float m})
0:12                  'anon@0' (layout( row_major std140) uniform block{ uniform 4-element array of 4-component vector of float a,  uniform 11-element array of structure{ temp 7-element array of 4-component vector of float m} s})
0:12                  Constant:
0:12                    1 (const uint)
0:12                'i' ( in int)
0:12              Constant:
0:12                0 (const int)
0:12            'i' ( in int)
0:10  Function Definition: PixelShaderFunction( ( temp void)
0:10    Function Parameters: 
0:?     Sequence
0:10      move second child to first child ( temp int)
0:?         'i' ( temp int)
0:?         'i' (layout( location=0) flat in int)
0:10      move second child to first child ( temp 3-element array of 4-component vector of float)
0:?         'input' ( temp 3-element array of 4-component vector of float)
0:?         'input' (layout( location=1) in 3-element array of 4-component vector of float)
0:10      move second child to first child ( temp 4-component vector of float)
0:?         '@entryPointOutput' (layout( location=0) out 4-component vector of float)
0:10        Function Call: @PixelShaderFunction(i1;vf4[3]; ( temp 4-component vector of float)
0:?           'i' ( temp int)
0:?           'input' ( temp 3-element array of 4-component vector of float)
0:?   Linker Objects
0:?     'anon@0' (layout( row_major std140) uniform block{ uniform 4-element array of 4-component vector of float a,  uniform 11-element array of structure{ temp 7-element array of 4-component vector of float m} s})
0:?     'C' ( global 4-component vector of float)
0:?     '@entryPointOutput' (layout( location=0) out 4-component vector of float)
0:?     'i' (layout( location=0) flat in int)
0:?     'input' (layout( location=1) in 3-element array of 4-component vector of float)

// Module Version 10000
// Generated by (magic number): 80001
// Id's are bound by 99

                              Capability Shader
               1:             ExtInstImport  "GLSL.std.450"
                              MemoryModel Logical GLSL450
                              EntryPoint Fragment 4  "PixelShaderFunction" 86 90 93
                              ExecutionMode 4 OriginUpperLeft
                              Source HLSL 500
                              Name 4  "PixelShaderFunction"
                              Name 17  "@PixelShaderFunction(i1;vf4[3];"
                              Name 15  "i"
                              Name 16  "input"
                              Name 20  "C"
                              Name 29  "b"
                              Name 45  ""
                              MemberName 45 0  "m"
                              Name 48  "$Global"
                              MemberName 48($Global) 0  "a"
                              MemberName 48($Global) 1  "s"
                              Name 50  ""
                              Name 84  "i"
                              Name 86  "i"
                              Name 88  "input"
                              Name 90  "input"
                              Name 93  "@entryPointOutput"
                              Name 94  "param"
                              Name 96  "param"
                              Decorate 42 ArrayStride 16
                              Decorate 44 ArrayStride 16
                              MemberDecorate 45 0 Offset 0
                              Decorate 47 ArrayStride 112
                              MemberDecorate 48($Global) 0 Offset 0
                              MemberDecorate 48($Global) 1 Offset 64
                              Decorate 48($Global) Block
                              Decorate 50 DescriptorSet 0
                              Decorate 86(i) Flat
                              Decorate 86(i) Location 0
                              Decorate 90(input) Location 1
                              Decorate 93(@entryPointOutput) Location 0
               2:             TypeVoid
               3:             TypeFunction 2
               6:             TypeInt 32 1
               7:             TypePointer Function 6(int)
               8:             TypeFloat 32
               9:             TypeVector 8(float) 4
              10:             TypeInt 32 0
              11:     10(int) Constant 3
              12:             TypeArray 9(fvec4) 11
              13:             TypePointer Function 12
              14:             TypeFunction 9(fvec4) 7(ptr) 13(ptr)
              19:             TypePointer Private 9(fvec4)
           20(C):     19(ptr) Variable Private
              21:    8(float) Constant 1065353216
              22:    8(float) Constant 1073741824
              23:    8(float) Constant 1077936128
              24:    8(float) Constant 1082130432
              25:    9(fvec4) ConstantComposite 21 22 23 24
              26:     10(int) Constant 10
              27:             TypeArray 9(fvec4) 26
              28:             TypePointer Function 27
              41:     10(int) Constant 4
              42:             TypeArray 9(fvec4) 41
              43:     10(int) Constant 7
              44:             TypeArray 9(fvec4) 43
              45:             TypeStruct 44
              46:     10(int) Constant 11
              47:             TypeArray 45(struct) 46
     48($Global):             TypeStruct 42 47
              49:             TypePointer Uniform 48($Global)
              50:     49(ptr) Variable Uniform
              51:      6(int) Constant 0
              52:      6(int) Constant 1
              53:             TypePointer Uniform 9(fvec4)
              60:      6(int) Constant 2
              61:             TypePointer Function 9(fvec4)
              69:      6(int) Constant 5
              85:             TypePointer Input 6(int)
           86(i):     85(ptr) Variable Input
              89:             TypePointer Input 12
       90(input):     89(ptr) Variable Input
              92:             TypePointer Output 9(fvec4)
93(@entryPointOutput):     92(ptr) Variable Output
4(PixelShaderFunction):           2 Function None 3
               5:             Label
           84(i):      7(ptr) Variable Function
       88(input):     13(ptr) Variable Function
       94(param):      7(ptr) Variable Function
       96(param):     13(ptr) Variable Function
                              Store 20(C) 25
              87:      6(int) Load 86(i)
                              Store 84(i) 87
              91:          12 Load 90(input)
                              Store 88(input) 91
              95:      6(int) Load 84(i)
                              Store 94(param) 95
              97:          12 Load 88(input)
                              Store 96(param) 97
              98:    9(fvec4) FunctionCall 17(@PixelShaderFunction(i1;vf4[3];) 94(param) 96(param)
                              Store 93(@entryPointOutput) 98
                              Return
                              FunctionEnd
17(@PixelShaderFunction(i1;vf4[3];):    9(fvec4) Function None 14
           15(i):      7(ptr) FunctionParameter
       16(input):     13(ptr) FunctionParameter
              18:             Label
           29(b):     28(ptr) Variable Function
              30:    9(fvec4) Load 20(C)
              31:    9(fvec4) Load 20(C)
              32:    9(fvec4) Load 20(C)
              33:    9(fvec4) Load 20(C)
              34:    9(fvec4) Load 20(C)
              35:    9(fvec4) Load 20(C)
              36:    9(fvec4) Load 20(C)
              37:    9(fvec4) Load 20(C)
              38:    9(fvec4) Load 20(C)
              39:    9(fvec4) Load 20(C)
              40:          27 CompositeConstruct 30 31 32 33 34 35 36 37 38 39
                              Store 29(b) 40
              54:     53(ptr) AccessChain 50 51 52
              55:    9(fvec4) Load 54
              56:      6(int) Load 15(i)
              57:     53(ptr) AccessChain 50 51 56
              58:    9(fvec4) Load 57
              59:    9(fvec4) FAdd 55 58
              62:     61(ptr) AccessChain 16(input) 60
              63:    9(fvec4) Load 62
              64:    9(fvec4) FAdd 59 63
              65:      6(int) Load 15(i)
              66:     61(ptr) AccessChain 16(input) 65
              67:    9(fvec4) Load 66
              68:    9(fvec4) FAdd 64 67
              70:     61(ptr) AccessChain 29(b) 69
              71:    9(fvec4) Load 70
              72:    9(fvec4) FAdd 68 71
              73:      6(int) Load 15(i)
              74:     61(ptr) AccessChain 29(b) 73
              75:    9(fvec4) Load 74
              76:    9(fvec4) FAdd 72 75
              77:      6(int) Load 15(i)
              78:      6(int) Load 15(i)
              79:     53(ptr) AccessChain 50 52 77 51 78
              80:    9(fvec4) Load 79
              81:    9(fvec4) FAdd 76 80
                              ReturnValue 81
                              FunctionEnd
